plugins/inputs/tencentcloudcm/tencentcloudcm.go

```diff
@@ -0,0 +1,371 @@
+				}
+				_, ok := t.discoverTool.registry[namespace.Name]
+				if len(region.Instances) == 0 && !ok {
+					return fmt.Errorf(
+						"unsupported namespace %s for discovering instances, please specify instances and dimensions",
+						namespace.Name,
+					)
+				}
+			}
+		}
+	}
+
+	// start discovery
+	go t.discoverTool.Discover(t.Accounts, t.DiscoveryInterval, t.Endpoint)
+	t.discoverTool.DiscoverMetrics()
+	for i := range t.Accounts {
+		for j := range t.Accounts[i].Namespaces {
+			// use discovered metrics if not specified
+			if len(t.Accounts[i].Namespaces[j].Metrics) == 0 {
+				metrics, ok := t.discoverTool.DiscoveredMetrics[t.Accounts[i].Namespaces[j].Name]
+				if !ok {
+					return fmt.Errorf("unsupported namespace %s for discovering metrics, please specify metrics",
+						t.Accounts[i].Namespaces[j].Name)
+				}
+				t.Accounts[i].Namespaces[j].Metrics = metrics
+			}
+		}
+	}
+
+	t.client = &cloudmonitorClient{Accounts: t.Accounts}
+
+	return nil
+
+}
+
+func (t *TencentCloudCM) updateWindow(relativeTo time.Time) {
+
+	windowEnd := relativeTo.Add(-time.Duration(t.Delay))
+
+	// starting point is two times the aggregation period to make sure all points are covered
+	t.windowStart = windowEnd.Add(-time.Duration(t.Period) * 2)
+
+	t.windowEnd = windowEnd
+}
+
+func (t *TencentCloudCM) Gather(acc telegraf.Accumulator) error {
+	t.updateWindow(time.Now())
+
+	metricObjects := t.client.GetMetricObjects(*t)
+
+	lmtr := limiter.NewRateLimiter(t.RateLimit, time.Second)
+	defer lmtr.Stop()
+
+	wg := sync.WaitGroup{}
+	rLock := sync.Mutex{}
+	results := []monitor.GetMonitorDataResponse{}
+
+	// requestIDMap contains request ID and metric objects for later aggregation
+	requestIDMap := map[string]metricObject{}
+	for _, obj := range metricObjects {
+
+		wg.Add(1)
+		<-lmtr.C
+		go func(m metricObject) {
+			defer wg.Done()
+
+			for {
+
+				client, err := t.client.NewClient(m.Region, m.Account.crs, *t)
+				if err != nil {
+					acc.AddError(err)
+					break
+				}
+
+				size := 100
+				if len(m.MonitorInstances) >= size {
+					batch := m.MonitorInstances[:size]
+					if len(batch) == 0 {
+						break
+					}
+
+					request := t.client.NewGetMonitorDataRequest(m.Namespace, m.Metric, batch, *t)
+
+					result, err := t.client.GatherMetrics(client, request, *t)
+					if err != nil {
+						acc.AddError(err)
+						break
+					}
+
+					rLock.Lock()
+					requestIDMap[*result.Response.RequestId] = m
+					results = append(results, *result)
+					rLock.Unlock()
+
+					m.MonitorInstances = m.MonitorInstances[size:]
+				} else {
+					request := t.client.NewGetMonitorDataRequest(m.Namespace, m.Metric, m.MonitorInstances, *t)
+
+					result, err := t.client.GatherMetrics(client, request, *t)
+					if err != nil {
+						acc.AddError(err)
+						break
+					}
+
+					rLock.Lock()
+					requestIDMap[*result.Response.RequestId] = m
+					results = append(results, *result)
+					rLock.Unlock()
+					break
+				}
```

Comment:
In this code you suffer terribly from a race condition as you access `m.MonitorInstances` in an unsafe fashion. If you start multiple parallel functions, `batch := m.MonitorInstances[:size]` might give you _exactly_ the same batch for most( or even all) of those functions as you remove that batch only late in the branch.

You at least need a lock protecting the access (and removal) of `m.MonitorInstances` to avoid those problems. However, if you do it naively, you will hold that lock for most of the time in that function. The question is then, why go for a parallel approach in the first place.
Alternatively, you need to correctly implement a _workerqueue_ or use an existing implementation.
However, please make sure that the gain of the parallel processing exceeds the complexity cost!

Reply:
Well, in current implementation, each `m.MonitorInstance` is a separate copy of all the instances that needs to be gathered within ONE `metricObject`. However within the goroutine, the `m.MonitorInstance` is not accessed in parallel. So it LGTM. 

Below is an illustration indicating that the slice will not be accessed in parallel. Please correct me if I am wrong.

```go
go func(m metricObject) {
    // do something

    // This slice will not be accessed in parallel as each slice is a copy within a different metricObject
    batch := m.MonitorInstances[:size]

    // do something
}(obj)
```

Reply:
You are right. I missed that `m` is passed in as a parameter filled by the for-loop. Please keep this without a lock (maybe add a short explanation to the function if you like).
However, my question about the gain from parallelism remains... Do you really gain much when doing the gathering in parallel? Given the increased complexity esp. in error-handling it is important to actually have a noticeable gain...

Reply:
The gain is noticeable when a product have a large number of metrics. 

Let's say we have `1000` CVM to monitor, each have `27` metrics (CPUUsage, MemUsage etc.). In current parallel implementation, all `27` metrics are pulled in parallel for `1000` CVM, which significantly boost the performance.

We could gather `27` metrics in above example in sequence instead of parallelism, but it would be much slower as every metric needs to wait for the predecessor metric to complete.

I totally agree that parallelism increased complexity. But from my point of view it might be necessary in terms of performance.

Reply:
OK, makes sense and is fine with me.